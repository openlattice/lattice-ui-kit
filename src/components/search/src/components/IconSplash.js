// @flow
import { Component } from 'react';
import type { ComponentType } from 'react';

import styled from 'styled-components';
import { IconDefinition } from '@fortawesome/fontawesome-svg-core';
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome';

import { NEUTRAL } from '../../../../colors';

const FigureWrapper = styled.figure`
  display: flex;
  flex-direction: column;
  justify-content: center;
  margin: 10px auto;
  text-align: center;

  > div:first-child {
    margin-bottom: 10px;
  }

  figcaption {
    color: ${NEUTRAL.N500};
    font-size: 16px;
    font-weight: 600;
    line-height: 22px;
    margin: 0 auto;
    text-align: center;
    width: 100%;
  }
`;

type Props = {
  className ? :string;
  /** Caption that renders underneath the icon */
  caption ? :string;
  /** Either a FontAwesome IconDefinition or a render prop */
  icon ? :IconDefinition | ComponentType<any>;
  /** size string that is passed to the FontAwesomeIcon or icon render prop */
  size ? :string;
}

class IconSplash extends Component<Props> {
  static defaultProps = {
    className: undefined,
    caption: '',
    icon: undefined,
    size: '5x'
  };

  renderIcon = () => {
    const { icon, size } = this.props;
    if (icon) {
      const content = typeof icon === 'function'
        ? icon(size)
        : <FontAwesomeIcon icon={icon} size={size} fixedWidth />;
      return <div>{content}</div>;
    }
    return null;
  }

  render() {
    const { className, caption } = this.props;
    return (
      <FigureWrapper className={className}>
        { this.renderIcon() }
        <figcaption>{caption}</figcaption>
      </FigureWrapper>
    );
  }
}

export default IconSplash;
