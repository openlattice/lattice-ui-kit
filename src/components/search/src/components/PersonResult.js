// @flow

import React, { Component } from 'react';
import styled from 'styled-components';
import { List, Map } from 'immutable';

import Label from '../../../../label';
import { Card } from '../../../../layout';
import {
  ResultGrid,
  ResultWrapper,
  ResultDetails,
  Truncated
} from './styled/StyledResultComponents';

const StyledResultGrid = styled(ResultGrid)`
  grid-template-columns: minmax(80px, 2fr) minmax(80px, 2fr) minmax(80px, 2fr) 1fr 1fr;

  > div:last-child {
    grid-column: 3 / -1;
  }
`;

type Props = {
  result ? :Map;
  resultLabels ? :Map;
}

class PersonResult extends Component<Props> {

  static defaultProps = {
    result: List(),
    resultLabels: Map()
  }

  transformResultToDetailsList = (result :Map) => {
    const { resultLabels } = this.props;
    const labels = result.map((value :any, key :string) => {
      let label = key;

      if (resultLabels && Map.isMap(resultLabels)) {
        label = resultLabels.get(key, key);
      }

      return Map({
        label,
        value,
        key
      });
    });
    return labels.toList();
  }

  render() {
    const { result } = this.props;
    const details :List<Map> = this.transformResultToDetailsList(result);

    return (
      <Card>
        <ResultWrapper>
          <ResultDetails>
            <StyledResultGrid>
              {
                details && details.map((detail :Map, index :number) => (
                  <div key={index.toString()}>
                    <Label bold>
                      {detail.get('label', '')}
                    </Label>
                    <Truncated>
                      {detail.get('value', '')}
                    </Truncated>
                  </div>
                ))
              }
            </StyledResultGrid>
          </ResultDetails>
        </ResultWrapper>
      </Card>
    );
  }
}

export default PersonResult;
export type { Props as ResultProps };
