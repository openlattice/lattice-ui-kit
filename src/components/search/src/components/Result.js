// @flow

import { Component } from 'react';
import isFunction from 'lodash/isFunction';
import { Map } from 'immutable';

import DataGrid from './DataGrid';
import { Card } from '../../../../layout';
import {
  ResultDetails,
  ResultWrapper,
} from './styled/StyledResultComponents';

type Props = {
  className ? :string;
  result :Map;
  resultColumns ? :number;
  resultLabels ? :Map;
  onClick ? :(result :Map) => void;
}

class Result extends Component<Props> {

  static defaultProps = {
    className: undefined,
    onClick: undefined,
    resultColumns: 4,
    resultLabels: undefined,
  }

  handleClick = () => {
    const { onClick, result } = this.props;
    if (isFunction(onClick)) {
      onClick(result);
    }
  }

  render() {
    const {
      className,
      result,
      resultColumns,
      resultLabels,
    } = this.props;

    return (
      <Card className={className} onClick={this.handleClick}>
        <ResultWrapper>
          <ResultDetails>
            <DataGrid
                columns={resultColumns}
                data={result}
                labelMap={resultLabels} />
          </ResultDetails>
        </ResultWrapper>
      </Card>
    );
  }
}

export default Result;
export type { Props as ResultProps };
