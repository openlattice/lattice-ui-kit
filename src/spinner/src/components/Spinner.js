import styled, { keyframes } from 'styled-components';

const spin = keyframes`
  0% {
    transform: rotate(0);
  }

  100% {
    transform: rotate(360deg);
  }
`;

const Spinner = styled.div`
  &::before {
    animation: ${spin} 0.75s linear infinite;
    border: 1px solid #c5d5e5;
    border-radius: 50%;
    border-top-color: #135;
    box-sizing: border-box;
    content: '';
    top: 50%;
    left: 50%;
    width: 50px;
    height: 50px;
    margin-top: -25px; /* half the height */
    margin-left: -25px; /* half the width */
    position: absolute;
  }
`;

export default Spinner;
