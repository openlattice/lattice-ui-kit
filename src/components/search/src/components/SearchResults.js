// @flow

import React, { Component } from 'react';
import type { ComponentType, Node } from 'react';
import { List, Map } from 'immutable';

import NotFound from './NotFound';
import Result from './Result';
import { CardStack } from '../../../../layout';

import type { ResultProps } from './Result';

type Props = {
  results :List<Map>;
  resultLabels ? :Map;
  resultComponent ? :ComponentType<ResultProps>;
  className ? :string;
  onResultClick ? :(result :Map) => void;
};

class SearchResults extends Component<Props> {

  static defaultProps = {
    className: undefined,
    onResultClick: undefined,
    resultComponent: Result,
    resultLabels: Map(),
  }

  renderResults = () :Node => {
    const {
      onResultClick,
      resultComponent: ResultComponent,
      resultLabels,
      results,
    } = this.props;
    if (List.isList(results) && results.count() && ResultComponent) {
      return results.map((result :Map, index :number) => (
        <ResultComponent
            key={index.toString()}
            onClick={onResultClick}
            result={result}
            resultLabels={resultLabels} />
      ));
    }

    return <NotFound caption="No results" />;
  }

  render() {
    const { className } = this.props;
    return (
      <CardStack className={className}>
        { this.renderResults() }
      </CardStack>
    );
  }
}

export default SearchResults;
export type { Props as SearchResultsProps };
