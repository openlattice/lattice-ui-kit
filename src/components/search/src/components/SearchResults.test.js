import React from 'react';
import toJson from 'enzyme-to-json';
import { mount, shallow } from 'enzyme';
import { List } from 'immutable';

import SearchResults from './SearchResults';
import PersonResult from './PersonResult';
import Result from './Result';
import {
  mockSearchResultsForPeople,
} from './constants';
import Spinner from '../../../../spinner';
import IconSplash from './IconSplash';

describe('SearchResults', () => {

  describe('snapshots', () => {
    test('should match snapshot', () => {
      const wrapper = mount(<SearchResults />);
      expect(toJson(wrapper)).toMatchSnapshot();
    });
  });

  describe('props', () => {

    test('should render custom ResultComponents', () => {
      const wrapper = mount(
        <SearchResults
            results={mockSearchResultsForPeople}
            resultComponent={PersonResult} />
      );
      expect(wrapper.find(PersonResult)).toHaveLength(3);
    });

    test('should render default Result when provided searchResults', () => {
      const wrapper = mount(<SearchResults results={mockSearchResultsForPeople} />);
      expect(wrapper.find(Result)).toHaveLength(3);
    });

    test('should render Spinner when isLoading', () => {
      const wrapper = shallow(<SearchResults isLoading />);
      expect(wrapper.find(Spinner)).toHaveLength(1);
    });

    test('should render default noResult when hasSearched and results are empty', () => {
      const wrapper = mount(<SearchResults hasSearched results={List()} />);
      expect(wrapper.find(IconSplash)).toHaveLength(1);
    });

    test('should render custom noResult when hasSearched and results are empty', () => {
      const customNoResults = () => (<div>I am custom</div>);
      const wrapper = mount(<SearchResults hasSearched results={List()} noResults={customNoResults} />);
      expect(wrapper.find(customNoResults)).toHaveLength(1);
    });

  });

});
