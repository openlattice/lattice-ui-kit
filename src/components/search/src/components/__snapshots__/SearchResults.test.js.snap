// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SearchResults snapshots should match snapshot 1`] = `
.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  position: relative;
}

.c0 > div {
  margin: 10px 0;
}

.c0 > div:first-child {
  margin-top: 0;
}

.c0 > div:last-child {
  margin-bottom: 0;
}

<SearchResults
  hasSearched={false}
  isLoading={false}
  noResults={[Function]}
  resultComponent={[Function]}
>
  <CardStack>
    <StyledComponent
      forwardedComponent={
        Object {
          "$$typeof": Symbol(react.forward_ref),
          "attrs": Array [],
          "componentStyle": ComponentStyle {
            "componentId": "CardStack-sc-1vtkaja-0",
            "isStatic": false,
            "lastClassName": "c0",
            "rules": Array [
              "display:flex;flex-direction:column;position:relative;& > div{margin:10px 0;}& > div:first-child{margin-top:0;}& > div:last-child{margin-bottom:0;}",
            ],
          },
          "displayName": "CardStack",
          "foldedComponentIds": Array [],
          "render": [Function],
          "styledComponentId": "CardStack-sc-1vtkaja-0",
          "target": "div",
          "toString": [Function],
          "warnTooManyClasses": [Function],
          "withComponent": [Function],
        }
      }
      forwardedRef={null}
    >
      <div
        className="c0"
      />
    </StyledComponent>
  </CardStack>
</SearchResults>
`;
